File: previous/20.valid-parentheses/20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
	let stack = [];

	for (let i = 0; i < s.length; i++) {
		if (s[i] == "(" || s[i] == "[" || s[i] == "{") {
			stack.push(s[i]);
			continue;
		}
		if (s[i] == ")") {
			if (stack.pop() != "(") return false;
		} else if (s[i] == "]") {
			if (stack.pop() != "[") return false;
		} else if (s[i] == "}") {
			if (stack.pop() != "{") return false;
		}
	}
	if (stack.length > 0) return false;
	return true;
};
// @lc code=end

File: 7-day-sprint/day4/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        paraDict = {'(': ')', '[': ']', '{': '}'}

        for p in s:
            if p in paraDict:
                stack.append(p)
            elif len(stack) == 0 or paraDict[stack.pop()] != p:
                return False

        return True and len(stack) == 0
# @lc code=end

# Time: O(n), Space: O(n)

# Concept: Stack

# Test Case
# "()[]{}"
# "(]"
# "([])"
# "([)]"
# "}{)("
# ")" 
# "("